src/node/node_id.h

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ds/json.h"
+
+#include <fmt/format.h>
+#include <string>
+
+namespace ccf
+{
+  struct NodeId
+  {
+    // The underlying value type should be blit-serialisable so that it can be
+    // written to and read from the ring buffer
+    using Value =
+      std::string; // < hex-encoded hash of node's identity public key
+    Value id;
+
+    NodeId() = default;
+
+    NodeId(const Value& id_) : id(id_) {}
+
+    void operator=(const NodeId& other)
+    {
+      id = other.id;
+    }
+
+    void operator=(const Value& id_)
+    {
+      id = id_;
+    }
+
+    bool operator==(const NodeId& other) const
+    {
+      return id == other.id;
+    }
+
+    bool operator!=(const NodeId& other) const
+    {
+      return !(*this == other);
+    }
+
+    bool operator<(const NodeId& other) const
+    {
+      return id < other.id;
+    }
+
+    operator Value() const
+    {
+      return id;
+    }
+
+    auto value() const
+    {
+      return id;
+    }
+
+    size_t size() const
+    {
+      return id.size();
+    }
+
+    MSGPACK_DEFINE(id);
+  };
+
+  inline void to_json(nlohmann::json& j, const NodeId& node_id)
+  {
+    j = node_id.id;
+  }
+
+  inline void from_json(const nlohmann::json& j, NodeId& node_id)
+  {
+    if (j.is_string())
+    {
+      node_id = j.get<std::string>();
+    }
+    else
+    {
+      throw std::runtime_error(
```

Comment:
Minor: If you throw a `JsonParseError` here, then the frontend might wrap it in a more helpful message (with a JSON pointer to where in the nested structure was wrong). This is undocumented and not used elsewhere though, so perhaps not worth relying on.